### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by fedorseliankin**
- [x] main page
- [x] details page
- [x] api integration
- [x] pull to refresh
screenshots:
Screenshot 2023-03-28 at 23 26 34

Screenshot 2023-03-28 at 23 27 18

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,13 +6,22 @@
     "android": "react-native run-android",
     "ios": "react-native run-ios",
     "lint": "eslint .",
+    "lint:fix": "eslint . --fix",
     "start": "react-native start",
     "test": "jest"
   },
   "dependencies": {
+    "@reduxjs/toolkit": "^1.9.3",
+    "@types/react-native-snap-carousel": "^3.8.5",
```
Comment (KunskayaV) [inline package.json]: dev deps

File: src/redux/types.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  attributes: {
+    name: string;
+    display_price: string | null;
+    price: string;
+    currency: string;
+    description: string | null;
+    display_compare_at_price: string | null;
+  };
+  relationships: {
+    variants: Array<object>;
```
Comment (KunskayaV) [inline src/redux/types.ts]: it's always better to use Record

File: src/redux/products.slice.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
+import {TProductResponse} from './types';
+
+export const productsSlice = createApi({
+  reducerPath: 'products',
+  baseQuery: fetchBaseQuery({baseUrl: ''}),
```
Comment (KunskayaV) [inline src/redux/products.slice.ts]: I believe u can use 'https://demo.spreecommerce.org/api' as baseUrl here

File: src/page/Details/Details.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
+import { useState, useCallback } from 'react';
+import {
+  Dimensions,
+  RefreshControl,
+  ScrollView,
+  StyleSheet,
+  Text,
+  TouchableOpacity,
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: Pls use Pressable instead. Just to get used to it. U can create common component with any feedback you wish and use it accross the app.

[It's the new API for handling basic touch events among others]
(https://reactnative.dev/blog/2020/07/06/version-0.63#pressable)

File: src/page/Details/Details.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    width: '100%',
+    flexWrap: 'wrap',
+    flexDirection: 'row',
+    marginTop: 20,
+    'productList:first-child': {
+      marginTop: 25,
+    },
+  },
+  title: {
+    fontFamily: 'System',
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: I believe this is default one and u can just skip it

File: src/page/Details/Details.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    fontStyle: 'normal',
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: same

File: src/page/Details/Details.tsx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    fontWeight: '400',
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: same

File: src/page/Details/Details.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    paddingHorizontal: 10,
+    paddingVertical: 5,
+  },
+  layout: {
+    height: '100%',
+    backgroundColor: 'white',
+  },
+});
+
+const AppButton = ({onPress, title}: {onPress?: () => void; title: string}) => (
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: It's better to have it as a separate component on app lvl (somewhere in 'src/components' folder

Just to not recreate it for each screen (btw, it's the 'screen' for RN, not the 'page')

File: src/page/Details/Details.tsx [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+export const Details = () => {
+  const {data} = useGetProductsQuery();
+  const product = data?.data[0];
+  const [refreshing, setRefreshing] = useState(false);
+
+  const onRefresh = useCallback(() => {
+    setRefreshing(true);
+    setTimeout(() => {
+      setRefreshing(false);
+    }, 2000);
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: pls use refetch instead of setting of timeout

File: src/page/Details/Details.tsx [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    {url: 'https://picsum.photos/250/250'},
+  ];
+  return (
+    <View style={styles.layout}>
+      <Header />
+      <ScrollView
+      refreshControl={
+        <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
+      }>  
+        <View style={styles.container}>
```
Comment (KunskayaV) [inline src/page/Details/Details.tsx]: I don't think you really need this container as `ScrollView` is a container itself. To style it use `contentContainerStyle` props

https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Components/ScrollView/ScrollView.js#L1731

File: src/page/Details/components/Carousel.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+});
+const renderItem = ({item}: {item: TImg}) => {
+  return (
+    <View>
+      <Image source={{uri: item.url}} style={styles.image} />
+    </View>
+  );
+};
+export const ImageCarousel = ({imgs}: TImageCarousel): JSX.Element => {
+  const isCarousel = useRef<Carousel<any>>(null);
```
Comment (KunskayaV) [inline src/page/Details/components/Carousel.tsx]: naming isCarousel has 'boolean' intension

File: src/page/Details/components/Carousel.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  const [index, setIndex] = useState(0);
+  const nextImg = () => isCarousel?.current?.snapToNext();
+  const prevImg = () => isCarousel?.current?.snapToPrev();
+  return (
+    <View style={styles.layout}>
+      <View style={styles.container}>
+        <View onTouchStart={prevImg}>
```
Comment (KunskayaV) [inline src/page/Details/components/Carousel.tsx]: If you want to handle touches use Pressable with onPressIn

File: src/page/Details/components/Carousel.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+          <Image
+            source={require('../../../assets/icons/Prev.png')}
```
Comment (KunskayaV) [inline src/page/Details/components/Carousel.tsx]: For icons it's better to .svg format.

https://github.com/kristerkari/react-native-svg-transformer
https://blog.openreplay.com/working-with-svgs-in-react-native/

File: src/page/Main/MainPage.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  return (
+    <View style={styles.container}>
+      <Header />
+      <SearchBar />
+      {data && (
+        <ScrollView
+        refreshControl={
+          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
+        }>  
+          <View style={styles.productList}>
```
Comment (KunskayaV) [inline src/page/Main/MainPage.tsx]: Same as for Details screen

File: src/page/Main/components/SearchBar/SearchBar.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+export const SearchBar = (): JSX.Element => {
+  return (
+    <View style={styles.container}>
+      <View style={styles.sectionStyle}>
+        <Image
+          source={require('../../../../assets/icons/Search.png')}
+          style={styles.imageStyle}
+        />
+        <TextInput
+          style={styles.textInput}
```
Comment (KunskayaV) [inline src/page/Main/components/SearchBar/SearchBar.tsx]: Don't remember whether search logic will be added in the following HW tasks but try to play with onChangeText to get familiar with TextInput API

File: package.json [hunk 16]
```diff
@@ -6,13 +6,22 @@
...
     "react": "18.2.0",
     "react-native": "0.71.4",
-    "react-native-flipper": "^0.184.0"
+    "react-native-flipper": "^0.184.0",
+    "react-native-gesture-handler": "^2.9.0",
+    "react-native-reanimated": "^3.0.2",
+    "react-native-reanimated-carousel": "^3.3.0",
+    "react-native-snap-carousel": "4.0.0-beta.6",
+    "react-native-snap-carousel-v4": "^1.0.1",
```
Comment (KunskayaV) [inline package.json:20]: do you really need 3 carousel packages?

File: src/page/Details/components/Carousel.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+        </View>
+        <View style={styles.carousel}>
+          <Carousel
+            vertical={false}
+            ref={isCarousel}
+            data={imgs}
+            renderItem={renderItem}
+            sliderWidth={SLIDER_WIDTH - 100}
+            itemWidth={ITEM_WIDTH}
+            onSnapToItem={ind => setIndex(ind)}
```
Comment (KunskayaV) [inline src/page/Details/components/Carousel.tsx]: I believe here can be simple {setIndex}

File: src/page/Main/MainPage.tsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    setTimeout(() => {
+      setRefreshing(false);
+    }, 2000);
+  }, []);
+  return (
+    <View style={styles.container}>
+      <Header />
+      <SearchBar />
+      {data && (
+        <ScrollView
```
Comment (KunskayaV) [inline src/page/Main/MainPage.tsx]: Not the best way to render list data

Pls use FlatList due to its performance optimization:
https://reactnative.dev/docs/flatlist
It has numColumns prop to render multiple columns
https://reactnative.dev/docs/flatlist#numcolumns

File: package.json (hunk 1 above)
Comment (fedorseliankin) [reply package.json]: fixed

File: src/redux/types.ts (hunk 2 above)
Comment (fedorseliankin) [reply src/redux/types.ts]: removed

File: src/redux/products.slice.ts (hunk 3 above)
Comment (fedorseliankin) [reply src/redux/products.slice.ts]: done

File: src/page/Details/Details.tsx (hunk 4 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 5 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 6 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 7 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 8 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 9 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/Details.tsx (hunk 10 above)
Comment (fedorseliankin) [reply src/page/Details/Details.tsx]: fixed

File: src/page/Details/components/Carousel.tsx (hunk 11 above)
Comment (fedorseliankin) [reply src/page/Details/components/Carousel.tsx]: fixed

File: src/page/Details/components/Carousel.tsx (hunk 12 above)
Comment (fedorseliankin) [reply src/page/Details/components/Carousel.tsx]: fixed

File: src/page/Details/components/Carousel.tsx (hunk 13 above)
Comment (fedorseliankin) [reply src/page/Details/components/Carousel.tsx]: fixed

File: src/page/Main/MainPage.tsx (hunk 14 above)
Comment (fedorseliankin) [reply src/page/Main/MainPage.tsx]: fixed

File: src/page/Main/components/SearchBar/SearchBar.tsx (hunk 15 above)
Comment (fedorseliankin) [reply src/page/Main/components/SearchBar/SearchBar.tsx]: add functionality for search and input

File: package.json (hunk 16 above)
Comment (fedorseliankin) [reply package.json:20]: remove redundant

File: src/page/Details/components/Carousel.tsx (hunk 17 above)
Comment (fedorseliankin) [reply src/page/Details/components/Carousel.tsx]: fixed

File: src/page/Main/MainPage.tsx (hunk 18 above)
Comment (fedorseliankin) [reply src/page/Main/MainPage.tsx]: fixed

File: metro.config.js [hunk 19]
```diff
@@ -1,17 +1,15 @@
-/**
- * Metro configuration for React Native
- * https://github.com/facebook/react-native
- *
- * @format
- */
-
-module.exports = {
-  transformer: {
-    getTransformOptions: async () => ({
```
Comment (KunskayaV) [inline metro.config.js:10]: pls leave this property for transformer as well

U can learn about metro configuration here - https://facebook.github.io/metro/docs/configuration/#gettransformoptions

File: src/screens/Details/components/Header/header.tsx [hunk 20]
```diff
@@ -0,0 +1,31 @@
+import {View, StyleSheet} from 'react-native';
+import BackArrow from '../../../../assets/icons/Arrow.svg';
```
Comment (KunskayaV) [inline src/screens/Details/components/Header/header.tsx:2]: can be useful

https://medium.com/@enzzorp/setting-up-aliases-in-react-native-de6c73b0dc5f

File: src/screens/Main/MainPage.tsx [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+      marginTop: 25,
+    },
+  },
+});
+export const MainPage = () => {
+  const [searchValue, onChangeSearchValue] = useState('');
+  const {data, isFetching, refetch} = useGetProductsQuery(searchValue);
+  const onRefresh = useCallback(() => {
+    refetch();
+  }, []);
```
Comment (KunskayaV) [inline src/screens/Main/MainPage.tsx:25]: missed dep: refetch

File: src/screens/Main/MainPage.tsx [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+  return (
+    <View style={styles.container}>
+      <Header />
+      <SearchBar searchValue={searchValue} onChange={onChangeSearchValue}/> 
+      <FlatList
+        contentContainerStyle={styles.productList}
+        refreshControl={<RefreshControl refreshing={isFetching} onRefresh={onRefresh} />}
+        data={data?.data}
+        keyExtractor={product => product.id}
```
Comment (KunskayaV) [inline src/screens/Main/MainPage.tsx:35]: 'id' is among the default ones. you can simply skip keyExtractor if you have ids in your data structure.

https://reactnative.dev/docs/flatlist#keyextractor

File: src/screens/Main/MainPage.tsx [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+        renderItem={({item}) => <Card product={item} />}
+        numColumns={2}
+        scrollEnabled={true}
```
Comment (KunskayaV) [inline src/screens/Main/MainPage.tsx:38]: default value is true. Can be omitted

File: src/screens/Main/MainPage.tsx [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+import { useCallback, useState } from 'react';
+
+const styles = StyleSheet.create({
+  container: {
+    height: '100%',
+    backgroundColor: 'white',
+  },
+  productList: {
+    marginTop: 20,
+    'productList:first-child': {
```
Comment (KunskayaV) [inline src/screens/Main/MainPage.tsx:15]: Don't really think it changes anything

**Review by KunskayaV — APPROVED** [review]
LGTM

Note: pay attention to metro configuration

**Pull request merged**